### ISSUE TITLE
bundle install --local does not install packaged gem from git source

### ISSUE THREAD

**Issue by chrobs** [bundler/bundler#3571]
Hi,

TLDR: I want to install a packaged set of gems in an offline environment but gems from git source are not installed (in contrast to the log output from `bundle install --local`)

For this purpose I'm packaging all gems on my local internet connected machine
with `bundle package --no-install --all`. Then I copy the `Gemfile` and
`Gemfile.lock` file as well as the `vendor` directory to the target machine and
run `bundle install --local --no-prune`.

All regular gems are installed properly but the gem from git source is left
(although bundle outputs that it can be used now).

Here is my test scenario setup:

Gemfile with one regular gem `curb` from rubygems.org and one gem `enc` from a
git source at github:
```bash
$ cat Gemfile
source 'https://rubygems.org'

gem 'curb', '0.8.8'
gem 'Enc',
  :require => 'enc/enc',
  :git => 'https://github.com/courts/enc.git',
  :ref => 'c6cb69c5e0fe4dc42793ceefe637644dc28304d7'
```

```bash
$ bundle package --no-install --all
Fetching https://github.com/courts/enc.git
Fetching gem metadata from https://rubygems.org/..
Fetching version metadata from https://rubygems.org/.
Resolving dependencies...
Using Enc 0.0.9 from https://github.com/courts/enc.git (at c6cb69c)
Installing curb 0.8.8
Using bundler 1.9.4
Bundle complete! 2 Gemfile dependencies, 3 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.
Updating files in vendor/cache
  * curb-0.8.8.gem
```

Bundler creates a `Gemfile.lock` and a `vendor/cache` directory:
```bash
$ cat Gemfile.lock
GIT
  remote: https://github.com/courts/enc.git
  revision: c6cb69c5e0fe4dc42793ceefe637644dc28304d7
  ref: c6cb69c5e0fe4dc42793ceefe637644dc28304d7
  specs:
    Enc (0.0.9)

GEM
  remote: https://rubygems.org/
  specs:
    curb (0.8.8)

PLATFORMS
  ruby

DEPENDENCIES
  Enc!
  curb (= 0.8.8)


$ tree vendor
vendor
└── cache
    ├── curb-0.8.8.gem
    └── enc-c6cb69c5e0fe
        ├── bin
        │   └── enc-cli
        ├── Enc.gemspec
        ├── lib
        │   └── enc
        │       ├── dec.rb
        │       └── enc.rb
        ├── Rakefile
        ├── README.markdown
        └── test
            ├── spec_dec.rb
            └── spec_enc.rb

6 directories, 9 files
```

Notice that the `enc`-Gem is not stored as a packed `.gem` file but as working
copy from the github repo instead.

(Now I have to delete the newly created `.bundle` directory because bundler has
saved the `no-install` option in there - anyway, is there any option to disable
this behaviour?)

No copy the files `Gemfile`, `Gemfile.lock` and the `vendor` directory to the
target machine (without internet).
The installed gems at the target machine at the moment:
```bash
$ gem list

*** LOCAL GEMS ***

bigdecimal (1.2.4)
bundler (1.9.4)
io-console (0.4.2)
json (1.8.1)
minitest (4.7.5)
psych (2.0.5)
rake (10.1.0)
rdoc (4.1.0)
test-unit (2.1.4.0)
```

Now running bundle install locally:

```bash
$ bundle install --local --no-prune
Using Enc 0.0.9 from https://github.com/courts/enc.git (at /tmp/fuzz/vendor/cache/enc-c6cb69c5e0fe)
Installing curb 0.8.8
Using bundler 1.9.4
Updating files in vendor/cache
Bundle complete! 2 Gemfile dependencies, 3 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.
bundle install --local --no-prune  36.72s user 5.16s system 95% cpu 44.062 total
```

The `curb` gem was installed properly but the `enc` gem is missing:

```bash
$ gem list

*** LOCAL GEMS ***

bigdecimal (1.2.4)
bundler (1.9.4)
curb (0.8.8)
io-console (0.4.2)
json (1.8.1)
minitest (4.7.5)
psych (2.0.5)
rake (10.1.0)
rdoc (4.1.0)
test-unit (2.1.4.0)
```

**Comment by indirect** [comment]
Have you tried this without passing `--no-install` to package? Does it work then?

You seem to be operating under the mistaken assumption that package will create a .gem file for git gems—that doesn't happen. However, this still looks like a bug, because `package --all` should then allow `install --local` to succeed. Can you please follow [ISSUES](https://github.com/carlhuda/bundler/blob/master/ISSUES.md) and then report back with the requested information?

**Comment by tempo** [comment]
Hi and thanks for your reply.

> Have you tried this without passing --no-install to package? Does it work then?

yap, but with same result.

> You seem to be operating under the mistaken assumption that package will create a .gem file for git gems—that doesn't happen

That was a wrong assumption on my side but should not concern the issue itself.

> However, this still looks like a bug, because package --all should then allow install --local to succeed

That's my point as well.


> Can you please follow ISSUES and then report back with the requested information?

I did, result does not changed. Here the further requested information:

```bash
$ bundle env
Environment

    Bundler   1.9.4
    Rubygems  2.2.2
    Ruby      2.1.4p265 (2014-10-27 revision 48166) [x86_64-linux]
    Git       2.1.4

Bundler settings

    cache_all
      Set for your local app (~/tmp/bundler-test/.bundle/config): "true"

Gemfile

    source 'https://rubygems.org'

    gem 'curb', '0.8.8'
    gem 'Enc',
      :require => 'enc/enc',
      :git => 'https://github.com/courts/enc.git',
      :ref => 'c6cb69c5e0fe4dc42793ceefe637644dc28304d7'

Gemfile.lock

    GIT
      remote: https://github.com/courts/enc.git
      revision: c6cb69c5e0fe4dc42793ceefe637644dc28304d7
      ref: c6cb69c5e0fe4dc42793ceefe637644dc28304d7
      specs:
        Enc (0.0.9)

    GEM
      remote: https://rubygems.org/
      specs:
        curb (0.8.8)

    PLATFORMS
      ruby

    DEPENDENCIES
      Enc!
      curb (= 0.8.8)
```

cheers!

**Comment by tempo** [comment]
bump

**Comment by indirect** [comment]
@tempo if this is important to you, feel free to send a pull request. Otherwise, we will fix this problem as soon as we are able to, and bumping it will not make a difference.

**Comment by tempo** [comment]
I'm sorry @indirect, annoying you was not my intention.
I was just missing some kind of feedback: does the provided information suffice, is the issue considered as a bug or are you just waiting for anything further.

**Comment by Envek** [comment]
It's pretty annoying for me too. I need to package whole app in Capistrano-style directory structure with `app/shared/bundle` path with gems and without  `app/current/vendor/cache`. Here is workaround:
```sh
# Temporary working copy is at ~/tmp/app and future production code is at ~/app/current
cd ~/tmp/app
bundle package --all
bundle install --deployment --local --path ~/app/shared/bundle --without development test
mkdir -p ~/app/shared/bundle/ruby/2.2.0/bundler/gems/
find ~/tmp/app/vendor/cache/* -maxdepth 0 -type d -exec cp -r "{}" ~/app/shared/bundle/ruby/2.2.0/bundler/gems/ \;
cp -r ~/tmp/app/.bundle ~/app/current/
```

**Comment by TimMoore** [comment]
I just read through this thread, but I don't understand what the bug is.

- Gems installed from git sources by Bundler _never_ appear in `gem list` (because `gem` does not support git gems)
- The `bundle install --local --no-prune` in the output pasted above appears to be successful

What am I missing?

**Comment by chrobs** [comment]
@TimMoore in short: you missed nothing. In fact, you are completely right.

My solution back in 2015 was to build a ruby gem from the git source, add the gem to the vendor directory and change the `Gemfile.lock` accordingly. I didn't had to solve a similar task in the meantime. Thanks to your comment, I finally got it right! The following is a short description of the problem and the solution to keep it documented for further use.

According to bundler's website (http://bundler.io/git.html):
> RubyGems lacks the ability to handle gems from git, any gems installed from a git repository will not show up in gem list.

Therefore, the listed invocations of `bundle package --no-install --all` `bundle install --local --no-prune` are successful.

To require a gem from git (installed via bundler) you have to require bundler first and then run `Bundler.setup` within your ruby script. The following IRB session illustrates this:

```
irb(main):001:0> require 'enc/enc'
LoadError: cannot load such file -- enc/enc
  from [...]/2.3.0/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
  from [...]/2.3.0/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
  from (irb):1
  from [...]/2.3.0/bin/irb:11:in `<main>'
irb(main):002:0> require 'bundler'
=> true
irb(main):003:0> Bundler.setup
=> "GIT\n  remote: https://github.com/courts/enc.git\n  revision: c6cb69c5e0fe4dc42793ceefe637644dc28304d7\n  ref: c6cb69c5e0fe4dc42793ceefe637644dc28304d7\n  specs:\n    Enc (0.0.9)\n\nGEM\n  remote: https://rubygems.org/\n  specs:\n    curb (0.8.8)\n\nPLATFORMS\n  ruby\n\nDEPENDENCIES\n  Enc!\n  curb (= 0.8.8)\n\nBUNDLED WITH\n   1.11.2\n"
irb(main):004:0> require 'enc/enc'
=> true
```
